Sources/Document/Document.swift

```diff
@@ -0,0 +1,219 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+class Document<T: JsonSpec> {
+    private var key: String
+    private var root: CRDTRoot
+    private var clone: CRDTRoot?
+    private var changeID: ChangeID
+    private var checkpoint: Checkpoint
+    private var localChanges: [Change]
+
+    init(key: String) {
+        self.key = key
+        self.root = CRDTRoot()
+        self.changeID = ChangeID.initial
+        self.checkpoint = Checkpoint.initial
+        self.localChanges = []
+    }
+
+    /**
+     * `update` executes the given updater to update this document.
+     */
+    func update(updater: (_ root: JsonObject<T>) -> Void, message: String? = nil) throws {
+        let clone = self.cloned()
+        let context = ChangeContext(id: self.changeID.next(), root: clone, message: message)
+
+        let proxy = JsonObject<T>(target: clone.getObject(), context: context)
+        updater(proxy)
+
+        if context.hasOperations() {
+            Logger.trivial("trying to update a local change: \(self.toJSON())")
+
+            let change = context.getChange()
+            try change.execute(root: self.root)
+            self.localChanges.append(change)
+            self.changeID = change.getID()
+
+            Logger.trivial("after update a local change: \(self.toJSON())")
+        }
+    }
+
+    /**
+     * `getCheckpoint` returns the checkpoint of this document.
+     *
+     */
+    func getCheckpoint() -> Checkpoint {
+        return self.checkpoint
+    }
+
+    /**
+     * `hasLocalChanges` returns whether this document has local changes or not.
+     *
+     */
+    func hasLocalChanges() -> Bool {
+        return self.localChanges.isEmpty == false
+    }
+
+    /**
+     * `ensureClone` make a clone of root.
+     */
+    func cloned() -> CRDTRoot {
+        if let clone = self.clone {
+            return clone
+        }
+
+        let clone = self.root.deepcopy()
+        self.clone = clone
+        return clone
+    }
+
+    /**
+     * `createChangePack` create change pack of the local changes to send to the
+     * remote server.
+     *
+     */
+    func createChangePack() -> ChangePack {
+        let changes = self.localChanges
+        let checkpoint = self.checkpoint.increasedClientSeq(by: changes.count)
+        return ChangePack(key: self.key, checkpoint: checkpoint, changes: changes)
+    }
+
+    /**
+     * `setActor` sets actor into this document. This is also applied in the local
+     * changes the document has.
+     *
+     */
+    func setActor(_ actorID: ActorID) {
+        for change in self.localChanges {
+            change.setActor(actorID)
+        }
+        self.changeID.setActor(actorID)
+
+        // TODOs also apply into root.
```

Comment:
What is the mean ```// TODO also apply into root.```?

Reply:
The issue below is detailed about this.
https://github.com/yorkie-team/yorkie/issues/417